
Контрольные вопросы:
1) В чём основная идея архитектуры трансформеров и чем она отличается от RNN/LSTM?

Основная идея трансформеров — это обработка всего входного текста (например, предложения) сразу, а не по частям последовательно, как в RNN/LSTM. Трансформер смотрит на все слова сразу и сам выбирает, какие слова важны для обработки каждого конкретного слова, используя механизм внимания (Attention).

Главные отличия от RNN/LSTM:
- Параллельность: Трансформеры обрабатывают все слова одновременно, что ускоряет обучение и позволяет работать с длинными текстами.
- Отсутствие рекуррентности: RNN/LSTM работают по шагам (один за другим), что сложнее и медленнее.
- Better long-range dependencies: Трансформеры легче улавливают связи между далекими словами, поскольку могут напрямую связать любое слово с любым другим.

2) Из каких основных блоков состоит трансформер (назовите ключевые компоненты)?

- Эмбеддинги (Embeddings): переводят слова в числовые векторы.
- Позиционные эмбеддинги (Positional embeddings): добавляют информацию о положении слова в последовательности, чтобы модель знала порядок слов.
- Self-Attention (Механизм само-внимания): позволяет модели увидеть, какие другие слова важны для каждого слова.
- Feed-forward сети (FFN): небольшие нейронные сети для «обработки» признаков после внимания.
- Нормализация и резидуальные связи (Layer norm + residual connections): улучшают стабильность и обучение.
- Множественные головы внимания (Multi-head Attention): позволяют модели смотреть на разные аспекты текста одновременно.
- Encoder и decoder: архитектура трансформера обычно состоит из encoder (для кодирования входа) и decoder (для генерации выхода), особенно в задачах перевода.

3) Для чего нужен механизм Attention и какие преимущества он даёт по сравнению с рекуррентными сетями?

Механизм Attention помогает модели сосредоточиться на наиболее важной части входных данных при обработке информации. 
Например, при переводе слова «банк» модель поймёт, что для правильного перевода важно учитывать контекст («река» или «деньги»).

Преимущества перед RNN:

Прямой доступ к любому слову последовательности: можно непосредственно учитывать важные слова, даже если они далеко.
Параллельная обработка: быстрее обучение и вычисления.
Лучшее понимание контекста: attention выделяет веса важности для каждого слова.

4) Какие задачи решают BERT и GPT, в чём их принципиальное отличие?

BERT (Bidirectional Encoder Representations from Transformers):

- Задача: понимание языка (например, классификация текста, ответ на вопросы, определение сущностей).
Особенность: двунаправленность, то есть модель учитывает контекст как слева, так и справа от слова.
Обучается на задачах заполнения пропусков в тексте (Masked Language Modeling) и предсказания следующего предложения.
GPT (Generative Pre-trained Transformer):

- Задача: генерация текста (например, создание связных абзацев, продолжение текста).
Особенность: однонаправленный — тексты генерируются слева направо, используя только предыдущий контекст.
Обучается на задаче предсказания следующего слова (Language Modeling).

BERT больше для понимания существующего текста, GPT — для генерации нового.

5) Что такое fine-tuning предобученной модели и почему он эффективен?

Fine-tuning — это когда мы берём уже обученную модель (например, BERT или GPT), которая знает много о языке, и дообучаем её на конкретной задаче (например, категоризация отзывов или анализ тональности).

Почему это эффективно?

- Модель уже научилась понимать язык, что экономит время и данные.
- Для конкретной задачи требуется немного обучения, так как базовое понимание уже есть.
- Повышается качество на целевой задаче без необходимости строить модель с нуля.

6) Как происходит токенизация в BERT (WordPiece)?

WordPiece токенизация разбивает слова на более мелкие части (субслова). Например, слово «playing» может разбиться на «play» + «##ing».

Если целое слово есть в словаре — используем его.
Если нет — разрезаем на части так, чтобы подчасти были известны модели.
Преимущество: модель может обработать редкие или новые слова, используя знакомые подслова.

7) Какие метрики применяются для оценки качества в задаче бинарной классификации текстов?

- Accuracy (Точность): доля правильно классифицированных примеров.
- Precision (Точность предсказаний положительного класса): из всех положительных предсказаний, сколько правильных.
- Recall (Полнота): сколько из всех реальных положительных примеров модель нашла.
- F1-score: гармоническое среднее Precision и Recall — баланс между ними.
- ROC-AUC: площадь под ROC-кривой — мера качества при разных порогах классификации.

8) Как проверить, «понимает» ли модель содержание текста, а не просто запоминает частые шаблоны?

1) Тестирование на новых данных с изменённым контекстом: если модель хорошо работает на примерах, отличающихся от обучающих.
2) Интерпретируемость модели: анализируем важность слов (например, с помощью attention или LIME/SHAP).
3) Adversarial examples: подаём чуть изменённые, но логичные тексты — если модель реагирует разумно, она действительно “понимает”.
4) Cross-domain evaluation: проверяем модель на данных из другой области.
Анализ ошибок: смотрим, где модель ошибается — если ошибки связаны с неверным контекстом, возможно, запоминание шаблонов.

Практические задания:
1) Увеличить размер выборки.
2) Сравнить с другой моделью.
3) Изменить гиперпараметры модели.
4) Обучить модели на русскоязычном наборе данных.

#1. Увеличение размера выборки:

In [ ]:
import numpy as np
import torch
import evaluate
from datasets import load_dataset
from transformers import BertTokenizerFast, BertForSequenceClassification, Trainer, TrainingArguments
from transformers import TrainingArguments


# 1. Загрузка набора данных IMDB (используем поднабор для демонстрации)
dataset = load_dataset("imdb")
small_train_dataset = dataset["train"].shuffle(seed=42)
small_test_dataset = dataset["test"].shuffle(seed=42)

# 2. Загрузка токенизатора
tokenizer = BertTokenizerFast.from_pretrained("bert-base-uncased")

# Функция токенизации
def tokenize(batch):
    return tokenizer(batch["text"], truncation=True, padding="max_length", max_length=128)

small_train_dataset = small_train_dataset.map(tokenize, batched=True)
small_test_dataset = small_test_dataset.map(tokenize, batched=True)

# Убираем лишние колонки
small_train_dataset = small_train_dataset.remove_columns(["text"])
small_test_dataset = small_test_dataset.remove_columns(["text"])

# 3. Загрузка модели для классификации
model = BertForSequenceClassification.from_pretrained("bert-base-uncased", num_labels=2)

# 4. Настройка параметров обучения
training_args = TrainingArguments(
    output_dir="./results",
    num_train_epochs=10,             
    per_device_train_batch_size=128,
    per_device_eval_batch_size=128,
    eval_strategy="epoch",
    logging_dir="./logs",
    logging_steps=50,
    save_strategy="epoch"
)

# Метрика для оценки
accuracy_metric = evaluate.load("accuracy")

def compute_metrics(eval_pred):
    logits, labels = eval_pred
    predictions = np.argmax(logits, axis=-1)
    return accuracy_metric.compute(predictions=predictions, references=labels)

# 5. Создание Trainer и запуск обучения
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=small_train_dataset,
    eval_dataset=small_test_dataset,
    compute_metrics=compute_metrics,
)

# Запуск дообучения
trainer.train()

# Оценка модели
results = trainer.evaluate()
print("Evaluation results:", results)

/home/aa-sablin/env_lab/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
2025-06-02 16:32:35.435172: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1748856755.450263    9315 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1748856755.454412    9315 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1748856755.467017    9315 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W000

При увеличении выборки можем заметить, что после второй эпохи наблюдается переобучение, так как потери на тренировочной выборке падают, а на валидационной — растут.

2) Сравнить с другой моделью.

In [1]:
import numpy as np
import torch
import evaluate
from datasets import load_dataset
from transformers import Trainer, TrainingArguments, DistilBertTokenizerFast, DistilBertForSequenceClassification

# 1. Загрузка набора данных IMDB (используем поднабор для демонстрации)
dataset = load_dataset("imdb")
small_train_dataset = dataset["train"].shuffle(seed=42).select(range(5000))
small_test_dataset = dataset["test"].shuffle(seed=42).select(range(1000))

# 2. Загрузка токенизатора
tokenizer = DistilBertTokenizerFast.from_pretrained("distilbert-base-uncased")

# Функция токенизации
def tokenize(batch):
    return tokenizer(batch["text"], truncation=True, padding="max_length", max_length=128)

small_train_dataset = small_train_dataset.map(tokenize, batched=True)
small_test_dataset = small_test_dataset.map(tokenize, batched=True)

# Убираем лишние колонки
small_train_dataset = small_train_dataset.remove_columns(["text"])
small_test_dataset = small_test_dataset.remove_columns(["text"])

# 3. Загрузка модели для классификации
model = DistilBertForSequenceClassification.from_pretrained("distilbert-base-uncased", num_labels=2)

# 4. Настройка параметров обучения
training_args = TrainingArguments(
    output_dir="./results",
    num_train_epochs=7,
    per_device_train_batch_size=64,
    per_device_eval_batch_size=64,
    eval_strategy="epoch",
    logging_dir="./logs",
    logging_steps=50,
    save_strategy="epoch"
)

# Метрика для оценки
accuracy_metric = evaluate.load("accuracy")

def compute_metrics(eval_pred):
    logits, labels = eval_pred
    predictions = np.argmax(logits, axis=-1)
    return accuracy_metric.compute(predictions=predictions, references=labels)

# 5. Создание Trainer и запуск обучения
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=small_train_dataset,
    eval_dataset=small_test_dataset,
    compute_metrics=compute_metrics,
)

# Запуск дообучения
trainer.train()

# Оценка модели
results = trainer.evaluate()
print("Evaluation results:", results)

/home/aa-sablin/env_lab/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
2025-06-02 17:54:07.041876: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1748861647.056563    6646 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1748861647.060665    6646 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1748861647.071114    6646 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W000

Evaluation results: {'eval_loss': 0.7626335620880127, 'eval_accuracy': 0.841, 'eval_runtime': 1.9357, 'eval_samples_per_second': 516.618, 'eval_steps_per_second': 8.266, 'epoch': 7.0}


DistilBERT демонстрирует оптимальное соотношение скорость-ресурсы (3.5 мин обучения, 84.1% точности), 
превосходя BERT-base по скорости в 4× при потреблении в 3× меньше памяти. BERT обеспечивает высокую точность (88.7%).

3) Изменить гиперпараметры модели.

In [5]:
import torch
import gc
import numpy as np
from transformers import DistilBertTokenizer, DistilBertForSequenceClassification, Trainer, TrainingArguments
from datasets import load_dataset
from sklearn.metrics import accuracy_score, f1_score

# Очистка памяти перед запуском
gc.collect()
torch.cuda.empty_cache()

# Фиксируем генератор случайных чисел для воспроизводимости
RANDOM_SEED = 42
np.random.seed(RANDOM_SEED)
torch.manual_seed(RANDOM_SEED)
if torch.cuda.is_available():
    torch.cuda.manual_seed_all(RANDOM_SEED)

# 1. Загрузка данных с буферизацией для ускорения доступа
dataset = load_dataset('imdb', keep_in_memory=True)
tokenizer = DistilBertTokenizer.from_pretrained('distilbert-base-uncased')

# 2. Предварительная токенизация с батчингом
def tokenize_batch(batch):
    return tokenizer(
        batch['text'],
        padding='max_length',
        truncation=True,
        max_length=192,  # Сокращенный размер для ускорения
        return_tensors='np'
    )

dataset = dataset.map(
    tokenize_batch,
    batched=True,
    batch_size=128,
    remove_columns=['text'],
    keep_in_memory=True
)

# Выбор компактного подмножества
train_dataset = dataset['train'].shuffle(seed=RANDOM_SEED).select(range(4000))
val_dataset = dataset['test'].shuffle(seed=RANDOM_SEED).select(range(800))

# 3. Функция метрик с буферизацией
def compute_metrics(p):
    predictions = np.argmax(p.predictions, axis=1)
    return {
        'accuracy': accuracy_score(p.label_ids, predictions),
        'f1': f1_score(p.label_ids, predictions, average='macro')
    }

# 4. Оптимизированный процесс поиска
learning_rates = [2e-5, 3e-5, 5e-5]
batch_sizes = [24, 32]
epoch_options = [3, 4]

results = []
best_acc = 0.0

for lr in learning_rates:
    for bs in batch_sizes:
        for epochs in epoch_options:
            gc.collect()
            torch.cuda.empty_cache()
            
            print(f"\n\033[1;34m||| LR={lr:.1e} | Batch={bs} | Epochs={epochs} \033[0m")
            
            model = DistilBertForSequenceClassification.from_pretrained(
                'distilbert-base-uncased',
                num_labels=2
            )
            
            args = TrainingArguments(
                output_dir='./hparam_tuning',
                learning_rate=float(lr),  # Явное преобразование во float
                per_device_train_batch_size=bs,
                per_device_eval_batch_size=bs*2,
                num_train_epochs=epochs,
                eval_strategy ='epoch',
                logging_strategy='epoch',
                weight_decay=0.01,
                fp16=True,  # Обязательное использование ускорения на GPU
                seed=RANDOM_SEED,
                save_strategy='no',
                optim='adamw_torch',
                logging_steps=50,
                disable_tqdm=True,
                report_to='none',
                log_level='warning'
            )
            
            trainer = Trainer(
                model=model,
                args=args,
                train_dataset=train_dataset,
                eval_dataset=val_dataset,
                compute_metrics=compute_metrics
            )
            
            trainer.train()
            metrics = trainer.evaluate()
            
            current_acc = metrics['eval_accuracy']
            results.append({
                'lr': lr,
                'batch_size': bs,
                'epochs': epochs,
                'accuracy': current_acc
            })
            
            if current_acc > best_acc:
                best_acc = current_acc
                best_config = results[-1].copy()
            
            print(f"► Accuracy: {current_acc:.4f} | " 
                  f"Memory: {torch.cuda.max_memory_allocated()/1024**2:.0f}MB")

# 5. Результаты
print("\n\033[1;92m" + "="*50)
print(f"⚡ Лучшая точность: {best_acc:.4f}")
print(f"⚙️ Оптимальные параметры:")
print(f"   • Скорость обучения: {best_config['lr']:.1e}")
print(f"   • Размер батча: {best_config['batch_size']}")
print(f"   • Эпохи: {best_config['epochs']}")
print("="*50 + "\033[0m")

print("\nДетальные результаты:")
for i, r in enumerate(results):
    print(f"Конфиг {i+1}: LR={r['lr']:.1e}, BS={r['batch_size']}, E={r['epochs']}"
          f" → Acc={r['accuracy']:.4f}")


Map: 100%|██████████| 50000/50000 [02:39<00:00, 314.25 examples/s]



||| LR=2.0e-05 | Batch=24 | Epochs=3 


Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight', 'pre_classifier.bias', 'pre_classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


{'loss': 0.4398, 'grad_norm': 8.28321647644043, 'learning_rate': 1.341317365269461e-05, 'epoch': 1.0}
{'eval_loss': 0.3193889856338501, 'eval_accuracy': 0.85875, 'eval_f1': 0.8587232898719914, 'eval_runtime': 1.0058, 'eval_samples_per_second': 795.387, 'eval_steps_per_second': 16.902, 'epoch': 1.0}
{'loss': 0.2398, 'grad_norm': 6.134963512420654, 'learning_rate': 6.746506986027944e-06, 'epoch': 2.0}
{'eval_loss': 0.3071049451828003, 'eval_accuracy': 0.87125, 'eval_f1': 0.8712497988278107, 'eval_runtime': 1.0113, 'eval_samples_per_second': 791.052, 'eval_steps_per_second': 16.81, 'epoch': 2.0}
{'loss': 0.1619, 'grad_norm': 9.876880645751953, 'learning_rate': 7.984031936127744e-08, 'epoch': 3.0}
{'eval_loss': 0.3346048593521118, 'eval_accuracy': 0.87625, 'eval_f1': 0.8762173136343816, 'eval_runtime': 1.0214, 'eval_samples_per_second': 783.231, 'eval_steps_per_second': 16.644, 'epoch': 3.0}
{'train_runtime': 58.431, 'train_samples_per_second': 205.371, 'train_steps_per_second': 8.574, 'tr

Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight', 'pre_classifier.bias', 'pre_classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


{'loss': 0.4414, 'grad_norm': 7.139427661895752, 'learning_rate': 1.5059880239520958e-05, 'epoch': 1.0}
{'eval_loss': 0.3162311911582947, 'eval_accuracy': 0.8575, 'eval_f1': 0.8574777308954524, 'eval_runtime': 1.0034, 'eval_samples_per_second': 797.282, 'eval_steps_per_second': 16.942, 'epoch': 1.0}
{'loss': 0.2379, 'grad_norm': 8.262118339538574, 'learning_rate': 1.0059880239520958e-05, 'epoch': 2.0}
{'eval_loss': 0.33282092213630676, 'eval_accuracy': 0.8675, 'eval_f1': 0.8673134094820842, 'eval_runtime': 1.0098, 'eval_samples_per_second': 792.233, 'eval_steps_per_second': 16.835, 'epoch': 2.0}
{'loss': 0.1535, 'grad_norm': 5.680226802825928, 'learning_rate': 5.0898203592814375e-06, 'epoch': 3.0}
{'eval_loss': 0.3611493706703186, 'eval_accuracy': 0.87875, 'eval_f1': 0.8785433338915749, 'eval_runtime': 1.0055, 'eval_samples_per_second': 795.599, 'eval_steps_per_second': 16.906, 'epoch': 3.0}
{'loss': 0.0993, 'grad_norm': 17.258590698242188, 'learning_rate': 8.982035928143712e-08, 'epoc

Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight', 'pre_classifier.bias', 'pre_classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


{'loss': 0.4418, 'grad_norm': 4.882176876068115, 'learning_rate': 1.3440000000000002e-05, 'epoch': 1.0}
{'eval_loss': 0.37021881341934204, 'eval_accuracy': 0.83875, 'eval_f1': 0.8378069997595394, 'eval_runtime': 0.9974, 'eval_samples_per_second': 802.059, 'eval_steps_per_second': 13.033, 'epoch': 1.0}
{'loss': 0.2566, 'grad_norm': 8.942131042480469, 'learning_rate': 6.773333333333334e-06, 'epoch': 2.0}
{'eval_loss': 0.3185597360134125, 'eval_accuracy': 0.87375, 'eval_f1': 0.873716653303763, 'eval_runtime': 0.9992, 'eval_samples_per_second': 800.64, 'eval_steps_per_second': 13.01, 'epoch': 2.0}
{'loss': 0.1768, 'grad_norm': 5.86332893371582, 'learning_rate': 1.0666666666666667e-07, 'epoch': 3.0}
{'eval_loss': 0.34154289960861206, 'eval_accuracy': 0.86625, 'eval_f1': 0.8662481191141751, 'eval_runtime': 0.999, 'eval_samples_per_second': 800.839, 'eval_steps_per_second': 13.014, 'epoch': 3.0}
{'train_runtime': 55.4455, 'train_samples_per_second': 216.429, 'train_steps_per_second': 6.763, '

Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight', 'pre_classifier.bias', 'pre_classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


{'loss': 0.4582, 'grad_norm': 4.014491081237793, 'learning_rate': 1.5080000000000001e-05, 'epoch': 1.0}
{'eval_loss': 0.3636631667613983, 'eval_accuracy': 0.84125, 'eval_f1': 0.8403818267797191, 'eval_runtime': 1.003, 'eval_samples_per_second': 797.638, 'eval_steps_per_second': 12.962, 'epoch': 1.0}
{'loss': 0.2486, 'grad_norm': 5.254928112030029, 'learning_rate': 1.008e-05, 'epoch': 2.0}
{'eval_loss': 0.33399537205696106, 'eval_accuracy': 0.86125, 'eval_f1': 0.8611143695014662, 'eval_runtime': 0.998, 'eval_samples_per_second': 801.57, 'eval_steps_per_second': 13.026, 'epoch': 2.0}
{'loss': 0.1705, 'grad_norm': 10.44433307647705, 'learning_rate': 5.0800000000000005e-06, 'epoch': 3.0}
{'eval_loss': 0.3487713634967804, 'eval_accuracy': 0.86375, 'eval_f1': 0.8637242353632484, 'eval_runtime': 0.9992, 'eval_samples_per_second': 800.663, 'eval_steps_per_second': 13.011, 'epoch': 3.0}
{'loss': 0.1159, 'grad_norm': 6.2729926109313965, 'learning_rate': 8e-08, 'epoch': 4.0}
{'eval_loss': 0.36902

Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight', 'pre_classifier.bias', 'pre_classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


{'loss': 0.4166, 'grad_norm': 11.150524139404297, 'learning_rate': 2.0119760479041917e-05, 'epoch': 1.0}
{'eval_loss': 0.32722190022468567, 'eval_accuracy': 0.8525, 'eval_f1': 0.8521664505540626, 'eval_runtime': 1.0159, 'eval_samples_per_second': 787.46, 'eval_steps_per_second': 16.734, 'epoch': 1.0}
{'loss': 0.2144, 'grad_norm': 7.79978084564209, 'learning_rate': 1.0179640718562873e-05, 'epoch': 2.0}
{'eval_loss': 0.36196762323379517, 'eval_accuracy': 0.87125, 'eval_f1': 0.8711031784642194, 'eval_runtime': 1.0287, 'eval_samples_per_second': 777.689, 'eval_steps_per_second': 16.526, 'epoch': 2.0}
{'loss': 0.1229, 'grad_norm': 6.063252925872803, 'learning_rate': 1.7964071856287427e-07, 'epoch': 3.0}
{'eval_loss': 0.40438881516456604, 'eval_accuracy': 0.865, 'eval_f1': 0.8649966249156229, 'eval_runtime': 1.0399, 'eval_samples_per_second': 769.287, 'eval_steps_per_second': 16.347, 'epoch': 3.0}
{'train_runtime': 58.5318, 'train_samples_per_second': 205.017, 'train_steps_per_second': 8.559

Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight', 'pre_classifier.bias', 'pre_classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


{'loss': 0.4264, 'grad_norm': 6.125901222229004, 'learning_rate': 2.2589820359281435e-05, 'epoch': 1.0}
{'eval_loss': 0.3075236976146698, 'eval_accuracy': 0.8625, 'eval_f1': 0.8624862486248626, 'eval_runtime': 1.0067, 'eval_samples_per_second': 794.711, 'eval_steps_per_second': 16.888, 'epoch': 1.0}
{'loss': 0.2214, 'grad_norm': 11.422225952148438, 'learning_rate': 1.5089820359281436e-05, 'epoch': 2.0}
{'eval_loss': 0.3654230535030365, 'eval_accuracy': 0.86875, 'eval_f1': 0.8685773023613843, 'eval_runtime': 1.0065, 'eval_samples_per_second': 794.81, 'eval_steps_per_second': 16.89, 'epoch': 2.0}
{'loss': 0.1249, 'grad_norm': 13.275642395019531, 'learning_rate': 7.634730538922157e-06, 'epoch': 3.0}
{'eval_loss': 0.39736407995224, 'eval_accuracy': 0.88, 'eval_f1': 0.8797565069265262, 'eval_runtime': 1.0068, 'eval_samples_per_second': 794.557, 'eval_steps_per_second': 16.884, 'epoch': 3.0}
{'loss': 0.0639, 'grad_norm': 27.2251033782959, 'learning_rate': 1.347305389221557e-07, 'epoch': 4.0}

Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight', 'pre_classifier.bias', 'pre_classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


{'loss': 0.4304, 'grad_norm': 4.959465980529785, 'learning_rate': 2.016e-05, 'epoch': 1.0}
{'eval_loss': 0.35503944754600525, 'eval_accuracy': 0.84375, 'eval_f1': 0.8429527397679784, 'eval_runtime': 1.0336, 'eval_samples_per_second': 773.982, 'eval_steps_per_second': 12.577, 'epoch': 1.0}
{'loss': 0.2303, 'grad_norm': 5.35066556930542, 'learning_rate': 1.0160000000000001e-05, 'epoch': 2.0}
{'eval_loss': 0.32905226945877075, 'eval_accuracy': 0.86875, 'eval_f1': 0.8687481542709194, 'eval_runtime': 1.0192, 'eval_samples_per_second': 784.966, 'eval_steps_per_second': 12.756, 'epoch': 2.0}
{'loss': 0.1369, 'grad_norm': 4.218013286590576, 'learning_rate': 1.6e-07, 'epoch': 3.0}
{'eval_loss': 0.3583962321281433, 'eval_accuracy': 0.865, 'eval_f1': 0.8649586435845977, 'eval_runtime': 1.0226, 'eval_samples_per_second': 782.346, 'eval_steps_per_second': 12.713, 'epoch': 3.0}
{'train_runtime': 55.4453, 'train_samples_per_second': 216.429, 'train_steps_per_second': 6.763, 'train_loss': 0.2658464762

Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight', 'pre_classifier.bias', 'pre_classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


{'loss': 0.4467, 'grad_norm': 4.346563339233398, 'learning_rate': 2.262e-05, 'epoch': 1.0}
{'eval_loss': 0.3548020124435425, 'eval_accuracy': 0.83625, 'eval_f1': 0.8353544827412851, 'eval_runtime': 1.0079, 'eval_samples_per_second': 793.748, 'eval_steps_per_second': 12.898, 'epoch': 1.0}
{'loss': 0.2282, 'grad_norm': 3.753711462020874, 'learning_rate': 1.5120000000000001e-05, 'epoch': 2.0}
{'eval_loss': 0.3418346643447876, 'eval_accuracy': 0.8625, 'eval_f1': 0.8624785122675418, 'eval_runtime': 1.0281, 'eval_samples_per_second': 778.099, 'eval_steps_per_second': 12.644, 'epoch': 2.0}
{'loss': 0.1349, 'grad_norm': 7.135671138763428, 'learning_rate': 7.62e-06, 'epoch': 3.0}
{'eval_loss': 0.3498421609401703, 'eval_accuracy': 0.875, 'eval_f1': 0.8748678291445339, 'eval_runtime': 1.019, 'eval_samples_per_second': 785.1, 'eval_steps_per_second': 12.758, 'epoch': 3.0}
{'loss': 0.0776, 'grad_norm': 4.697340965270996, 'learning_rate': 1.2000000000000002e-07, 'epoch': 4.0}
{'eval_loss': 0.4072668

Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight', 'pre_classifier.bias', 'pre_classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


{'loss': 0.4172, 'grad_norm': 9.662147521972656, 'learning_rate': 3.3532934131736525e-05, 'epoch': 1.0}
{'eval_loss': 0.32959669828414917, 'eval_accuracy': 0.84875, 'eval_f1': 0.8482699165327795, 'eval_runtime': 1.0062, 'eval_samples_per_second': 795.035, 'eval_steps_per_second': 16.894, 'epoch': 1.0}
{'loss': 0.1884, 'grad_norm': 2.684006929397583, 'learning_rate': 1.6966067864271457e-05, 'epoch': 2.0}
{'eval_loss': 0.44730690121650696, 'eval_accuracy': 0.86375, 'eval_f1': 0.8636372876956109, 'eval_runtime': 1.0195, 'eval_samples_per_second': 784.729, 'eval_steps_per_second': 16.675, 'epoch': 2.0}
{'loss': 0.0832, 'grad_norm': 4.819431781768799, 'learning_rate': 2.9940119760479047e-07, 'epoch': 3.0}
{'eval_loss': 0.523221492767334, 'eval_accuracy': 0.86875, 'eval_f1': 0.8687497949215546, 'eval_runtime': 1.002, 'eval_samples_per_second': 798.403, 'eval_steps_per_second': 16.966, 'epoch': 3.0}
{'train_runtime': 58.4433, 'train_samples_per_second': 205.327, 'train_steps_per_second': 8.57

Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight', 'pre_classifier.bias', 'pre_classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


{'loss': 0.4133, 'grad_norm': 5.330784320831299, 'learning_rate': 3.7724550898203595e-05, 'epoch': 1.0}
{'eval_loss': 0.3450813591480255, 'eval_accuracy': 0.8425, 'eval_f1': 0.8417244498040398, 'eval_runtime': 1.029, 'eval_samples_per_second': 777.466, 'eval_steps_per_second': 16.521, 'epoch': 1.0}
{'loss': 0.2088, 'grad_norm': 13.03296184539795, 'learning_rate': 2.5224550898203592e-05, 'epoch': 2.0}
{'eval_loss': 0.3902049958705902, 'eval_accuracy': 0.86375, 'eval_f1': 0.8637395675606414, 'eval_runtime': 1.0436, 'eval_samples_per_second': 766.587, 'eval_steps_per_second': 16.29, 'epoch': 2.0}
{'loss': 0.0992, 'grad_norm': 21.583200454711914, 'learning_rate': 1.2724550898203593e-05, 'epoch': 3.0}
{'eval_loss': 0.4874677360057831, 'eval_accuracy': 0.86375, 'eval_f1': 0.8637480839574306, 'eval_runtime': 1.0011, 'eval_samples_per_second': 799.16, 'eval_steps_per_second': 16.982, 'epoch': 3.0}
{'loss': 0.0414, 'grad_norm': 0.05090377852320671, 'learning_rate': 2.2455089820359282e-07, 'epoc

Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight', 'pre_classifier.bias', 'pre_classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


{'loss': 0.411, 'grad_norm': 3.5291192531585693, 'learning_rate': 3.3600000000000004e-05, 'epoch': 1.0}
{'eval_loss': 0.31074029207229614, 'eval_accuracy': 0.855, 'eval_f1': 0.8547376198258103, 'eval_runtime': 0.9985, 'eval_samples_per_second': 801.208, 'eval_steps_per_second': 13.02, 'epoch': 1.0}
{'loss': 0.1893, 'grad_norm': 2.9332337379455566, 'learning_rate': 1.6933333333333333e-05, 'epoch': 2.0}
{'eval_loss': 0.34411078691482544, 'eval_accuracy': 0.86875, 'eval_f1': 0.8687481542709194, 'eval_runtime': 0.9582, 'eval_samples_per_second': 834.859, 'eval_steps_per_second': 13.566, 'epoch': 2.0}
{'loss': 0.0937, 'grad_norm': 4.812600135803223, 'learning_rate': 2.6666666666666667e-07, 'epoch': 3.0}
{'eval_loss': 0.4434668719768524, 'eval_accuracy': 0.86875, 'eval_f1': 0.8687399504024527, 'eval_runtime': 0.9563, 'eval_samples_per_second': 836.521, 'eval_steps_per_second': 13.593, 'epoch': 3.0}
{'train_runtime': 54.1649, 'train_samples_per_second': 221.546, 'train_steps_per_second': 6.92

Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight', 'pre_classifier.bias', 'pre_classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


{'loss': 0.429, 'grad_norm': 3.9998083114624023, 'learning_rate': 3.77e-05, 'epoch': 1.0}
{'eval_loss': 0.35855910181999207, 'eval_accuracy': 0.83125, 'eval_f1': 0.8299102776714717, 'eval_runtime': 0.9581, 'eval_samples_per_second': 834.948, 'eval_steps_per_second': 13.568, 'epoch': 1.0}
{'loss': 0.2051, 'grad_norm': 5.5343403816223145, 'learning_rate': 2.5200000000000003e-05, 'epoch': 2.0}
{'eval_loss': 0.33785584568977356, 'eval_accuracy': 0.8625, 'eval_f1': 0.8623959369273013, 'eval_runtime': 0.9577, 'eval_samples_per_second': 835.371, 'eval_steps_per_second': 13.575, 'epoch': 2.0}
{'loss': 0.0949, 'grad_norm': 7.714051723480225, 'learning_rate': 1.27e-05, 'epoch': 3.0}
{'eval_loss': 0.5267953276634216, 'eval_accuracy': 0.845, 'eval_f1': 0.8449844984498449, 'eval_runtime': 0.9576, 'eval_samples_per_second': 835.452, 'eval_steps_per_second': 13.576, 'epoch': 3.0}
{'loss': 0.0419, 'grad_norm': 0.16962431371212006, 'learning_rate': 2.0000000000000002e-07, 'epoch': 4.0}
{'eval_loss': 0.

4) Обучить модели на русскоязычном наборе данных.

In [6]:
import transformers
print(transformers.__version__)


4.52.3


In [2]:
import torch
print(torch.cuda.is_available())  # Должно вывести True, если GPU доступна
print(torch.cuda.current_device())
print(torch.cuda.get_device_name(torch.cuda.current_device()))


True
0
NVIDIA GeForce RTX 3060


In [2]:
import pandas as pd

df = pd.read_csv("kt.csv")
df

Unnamed: 0                                             review  \
0              0  Фильм сняли просто чудно. Disney на этот раз н...   
1              1  На самом деле, к фильму у меня было предубежде...   
2              2  Пожалуй нет ни одного увлеченного киномана, кт...   
3              3  Я обожаю старый мультфильм 'Красавица и чудови...   
4              4  Никогда не фанател от «Форсажа», для меня он в...   
...          ...                                                ...   
2995        2995  Когда в твоих руках философский камень, очень ...   
2996        2996  Фильм не плохой, но и не достаточно хороший. О...   
2997        2997  Вот что называется фильм продуман до мелочей! ...   
2998        2998  Знаете, когда речь заходит об этом фильме, то ...   
2999        2999  Благодаря рецензиям ряда пользователей 'Кинопо...   

                                            translation     type  
0     The film was shot just wonderfully. Disney thi...     good  
1     In fact, I had a bias towards the film. I thou...     good  
2     Perhaps there is not a single keen film fan wh...     good  
3     I love the old Beauty and the Beast cartoon. L...     good  
4     I have never been a fan of Fast and the Furiou...     good  
...                                                 ...      ...  
2995  When the philosopher's stone is in your hands,...  neutral  
2996  The film is not bad, but not good enough eithe...  neutral  
2997  That's what the film is called thought out to ...  neutral  
2998  You know, when it comes to this film, I immedi...  neutral  
2999  Thanks to the reviews of a number of users of ...  neutral  

[3000 rows x 4 columns]

In [3]:
df.info()
df["type"].unique()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Unnamed: 0   3000 non-null   int64 
 1   review       3000 non-null   object
 2   translation  3000 non-null   object
 3   type         3000 non-null   object
dtypes: int64(1), object(3)
memory usage: 93.9+ KB


array(['good', 'bad', 'neutral'], dtype=object)

In [6]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from datasets import Dataset
from transformers import (
    BertTokenizerFast,
    BertForSequenceClassification,
    Trainer,
    TrainingArguments,
    EarlyStoppingCallback
)
import evaluate
import torch


# Преобразование меток в числовой формат
label_map = {'good': 2, 'neutral': 1, 'bad': 0}
df['label'] = df['type'].map(label_map).astype(int)

# Проверка распределения классов
print("Распределение классов:")
print(df['type'].value_counts())

# Разделение на train/val/test
train_df, temp_df = train_test_split(df, test_size=0.3, random_state=42, stratify=df['label'])
val_df, test_df = train_test_split(temp_df, test_size=0.5, random_state=42, stratify=temp_df['label'])

# Конвертация в формат Dataset
train_dataset = Dataset.from_pandas(train_df[['review', 'label']].reset_index(drop=True))
val_dataset = Dataset.from_pandas(val_df[['review', 'label']].reset_index(drop=True))
test_dataset = Dataset.from_pandas(test_df[['review', 'label']].reset_index(drop=True))

# 2. Инициализация токенизатора
tokenizer = BertTokenizerFast.from_pretrained('bert-base-multilingual-cased')

# Функция токенизации
def tokenize_function(batch):
    return tokenizer(batch['review'], truncation=True, padding='max_length', max_length=128)

# Применение токенизации
train_dataset = train_dataset.map(tokenize_function, batched=True)
val_dataset = val_dataset.map(tokenize_function, batched=True)
test_dataset = test_dataset.map(tokenize_function, batched=True)

# Удаление лишних колонок
train_dataset = train_dataset.remove_columns(['review'])
val_dataset = val_dataset.remove_columns(['review'])
test_dataset = test_dataset.remove_columns(['review'])

# 3. Загрузка модели
model = BertForSequenceClassification.from_pretrained(
    'bert-base-multilingual-cased',
    num_labels=3,  # Теперь у нас 3 класса
    problem_type="single_label_classification",
    id2label={0: "bad", 1: "neutral", 2: "good"},
    label2id={"bad": 0, "neutral": 1, "good": 2}
)

# 4. Настройка обучения
training_args = TrainingArguments(
    output_dir='./results',
    eval_strategy='epoch',
    save_strategy='epoch',
    learning_rate=2.0e-05,
    per_device_train_batch_size=24,
    per_device_eval_batch_size=24,
    num_train_epochs=8,
    weight_decay=0.01,
    load_best_model_at_end=True,
    metric_for_best_model='f1',
    logging_dir='./logs',
    logging_steps=50,
    report_to='none',
    remove_unused_columns=True
)

# Метрики
accuracy_metric = evaluate.load('accuracy')
f1_metric = evaluate.load('f1')

def compute_metrics(eval_pred):
    logits, labels = eval_pred
    predictions = np.argmax(logits, axis=-1)
    accuracy = accuracy_metric.compute(predictions=predictions, references=labels)['accuracy']
    f1 = f1_metric.compute(predictions=predictions, references=labels, average='weighted')['f1']
    return {'accuracy': accuracy, 'f1': f1}

# 5. Создание Trainer
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=val_dataset,
    compute_metrics=compute_metrics,
    callbacks=[EarlyStoppingCallback(early_stopping_patience=2)]
)

# Обучение
trainer.train()

print("\nОценка на тестовых данных...")
test_results = trainer.evaluate(test_dataset)
print("\nTest Results:")
print(f"Accuracy: {test_results['eval_accuracy']:.4f}")
print(f"F1 Score: {test_results['eval_f1']:.4f}")

Распределение классов:
type
good       1001
neutral    1000
bad         999
Name: count, dtype: int64


Map: 100%|██████████| 450/450 [00:00<00:00, 2575.92 examples/s]
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-multilingual-cased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.



Оценка на тестовых данных...



Test Results:
Accuracy: 0.4667
F1 Score: 0.4637


Модель тренируется, но не достигает удовлетворительной производительности даже на обучающих данных. 
Текущие результаты (task accuracy ~52% validation / ~47% test) недостаточны для практического использования, 
так как близка к базовому уровню (случайное угадывание). 